node/src/services/builder.rs

```diff
@@ -0,0 +1,166 @@
+            file_transfer_request_protocol_name,
+            network,
+        )
+        .await;
+
+        self.file_transfer = Some(file_transfer_service_handle);
+        self
+    }
+
+    pub async fn with_blockchain(
+        &mut self,
+        client: Arc<ParachainClient>,
+        rpc_handlers: Arc<RpcHandlers>,
+        keystore: KeystorePtr,
+    ) -> &mut Self {
+        let blockchain_service_handle = spawn_blockchain_service(
+            &self
+                .task_spawner
+                .as_ref()
+                .expect("Task spawner is not set."),
+            client.clone(),
+            rpc_handlers.clone(),
+            keystore.clone(),
+        )
+        .await;
+
+        self.blockchain = Some(blockchain_service_handle);
+        self
+    }
+
+    pub fn with_file_storage(&mut self, file_storage: Arc<RwLock<FL>>) -> &mut Self {
+        self.file_storage = Some(file_storage);
+        self
+    }
+
+    pub fn with_forest_storage(&mut self, forest_storage: Arc<RwLock<FS>>) -> &mut Self {
+        self.forest_storage = Some(forest_storage);
+        self
+    }
+
+    pub fn with_provider_pub_key(&mut self, provider_pub_key: [u8; 32]) -> &mut Self {
+        self.provider_pub_key = Some(provider_pub_key);
+        self
+    }
+
+    pub fn _file_storage(&self) -> &Option<Arc<RwLock<FL>>> {
+        &self.file_storage
+    }
+
+    pub fn build(self) -> StorageHubHandler<T, FL, FS> {
+        StorageHubHandler::<T, FL, FS>::new(
+            self.task_spawner.expect("Task Spawner not set"),
+            self.file_transfer.expect("File Transfer not set."),
+            self.blockchain.expect("Blockchain Service not set."),
+            self.file_storage.expect("File Storage not set."),
+            self.forest_storage.expect("Forest Storage not set."),
+        )
+    }
+}
+
+/// Provides an interface for defining the concrete types of
+/// each `StorageLayer` kind, so that their specific requirements can be fulfilled.
+pub trait StorageLayerBuilder {
+    fn setup_storage_layer(&mut self) -> &mut Self;
+}
+
+impl<T> StorageLayerBuilder
+    for StorageHubBuilder<T, InMemoryFileStorage<T>, InMemoryForestStorage<T>>
+where
+    T: TrieLayout + Send + Sync,
+    HasherOutT<T>: TryFrom<[u8; 32]>,
+{
+    fn setup_storage_layer(&mut self) -> &mut Self {
+        self.with_file_storage(Arc::new(RwLock::new(InMemoryFileStorage::<T>::new())))
+            .with_forest_storage(Arc::new(RwLock::new(InMemoryForestStorage::<T>::new())))
+    }
+}
+
+// TODO: Change this to RocksDB File Storage once it is implemented.
+impl<T> StorageLayerBuilder
+    for StorageHubBuilder<T, InMemoryFileStorage<T>, RocksDBForestStorage<T>>
+where
+    T: TrieLayout + Send + Sync,
+    HasherOutT<T>: TryFrom<[u8; 32]>,
+{
+    fn setup_storage_layer(&mut self) -> &mut Self {
+        let provider_pub_key = self
+            .provider_pub_key
+            .expect("Provider public key not set before building the storage layer.");
```

Comment:
This is not necessary, but with builder patterns you could instead have a function called `as_provider` that returns another struct builder called `StorageHubProviderBuilder` and this where the struct would be implemented. 

So the `setup_storage_layer` is only callable once you have the appropriate builder, and from there you can execute `with_provider_pub_key` which would also be moved to this builder. Finally execute `build()` which would execute `setup_storage_layer()` for you since being a provider requires a storage layer no matter what.

Reply:
It also separates the layers of services we expose with the builders to be available only for specific needs.

Reply:
I see your point. However setting up the storage layer is one of the first things we do, not the last. So that conversion of two different `Builders` would kind of be wasted I think.